test/metric_test.cc

```diff
@@ -0,0 +1,130 @@
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "snap/config.h"
+#include "snap/metric.h"
+#include "gtest/gtest.h"
+
+#include <chrono>
+#include <ctime>
+#include <sstream>
+#include <string>
+#include <vector>
+
+using std::chrono::system_clock;
+using std::make_pair;
+using std::pair;
+using std::string;
+using Plugin::Metric;
+using Plugin::ConfigPolicy;
+using Plugin::StringRule;
+
+
+static string extract_ns(const Metric& metric) {
+    string ns_str;
+    for (auto const& elem : metric.ns()) {
+        ns_str += "/" + elem.value;
+    }
+    return ns_str;
+}
+
+TEST(MetricTest, SetNsUnitDescriptionWorks) {
+    Metric fake_metric{
+            {
+                    {"foo", "", ""},
+                    {"bar", "", ""},
+            },
+            "atoms",
+            "critical metric"
+    };
+
+    EXPECT_EQ("/foo/bar", extract_ns(fake_metric));
+    EXPECT_EQ("atoms", fake_metric.get_rpc_metric_ptr()->unit());
+    EXPECT_EQ("critical metric", fake_metric.get_rpc_metric_ptr()->description());
+}
+
+TEST(MetricTest, SetNsWorks) {
+    Metric fake_metric;
+    fake_metric.set_ns({
+                               {"foo", "", ""},
+                               {"cluster", "", ""},
+                               {"node_count", "", ""}
+                       });
+    EXPECT_EQ("/foo/cluster/node_count", extract_ns(fake_metric));
+}
+
+TEST(MetricTest, SetFromRpcMetricWorks) {
+    Metric base_metric{
+            {
+                    {"foo", "", ""},
+                    {"bar", "", ""},
+            },
+            "atoms",
+            "critical metric"
+    };
+
+    rpc::Metric source_metric(*base_metric.get_rpc_metric_ptr());
+    Metric fake_metric(&source_metric);
+    EXPECT_EQ("/foo/bar", extract_ns(fake_metric));
+    EXPECT_EQ("atoms", fake_metric.get_rpc_metric_ptr()->unit());
+    EXPECT_EQ("critical metric", fake_metric.get_rpc_metric_ptr()->description());
+}
+
+TEST(MetricTest, SetTagsWorks) {
+    Metric fake_metric;
+    fake_metric.add_tag(make_pair("host", "zero"));
+    fake_metric.add_tag(make_pair("period", "1hr"));
+    EXPECT_EQ("zero", fake_metric.tags().at("host"));
+    EXPECT_EQ("1hr", fake_metric.tags().at("period"));
+}
+
+TEST(MetricTest, SetTimestampWorks) {
+    Metric fake_metric;
+    std::tm source_time{56,10,8,2,5,92,6,122,1}; // 02 May 1992, 08:10:56
+    std::time_t raw_time = mktime(&source_time);
+    fake_metric.set_timestamp(system_clock::from_time_t(raw_time));
+    std::time_t metric_rawtime = system_clock::to_time_t(fake_metric.timestamp());
+    tm metric_time = *gmtime(&metric_rawtime);
+    EXPECT_EQ(92, metric_time.tm_year);
+    EXPECT_EQ(5, metric_time.tm_mon);
+    EXPECT_EQ(2, metric_time.tm_mday);
+}
```

Comment:
Not for nanoseconds, as the resolution of the test time object doesn't allow this.